### ISSUE TITLE
Is this really seq2seq?

### ISSUE THREAD

**Issue by unrealwill** [bstriner/keras-seq2seq#1]
Hello, (unrealwill from keras Seq2seq thread) (not bouncing on the Keras Seq2seq discussion to avoid being rude to you)

I had a quick look at your code. 
The details look nice.
But it doesn't seem to be a seq2seq architecture.
I don't see the encoder-decoder structure, and you fall for the beginner's trap of not implementing teachers forcing.

Please read Graves again.

**Comment by bstriner** [comment]
Hi @unrealwill . It might not be the Seq2Seq you are looking for but it is definitely a Seq2Seq architecture. It isn't particularly deep. If you want to put Memory networks or something else inside, go for it.

The encoder and decoder share the hidden transition function, except the encoder receives real input and the decoder receives generated input. The softmax output and sampling is only part of the decoder network. Think of it as the unit is the encoder or the decoder depending on the value of the mask.

If you want the encoder and decoder to have different transition functions, you can put an extra switch inside the layer to use different functions in different modes. You could also break them into separate layers. I just wanted to see what happens when they share parameters, and it works.

In input mode and training mode, the real input and correct output are used and each softmax output is trained. During testing mode, the output is looped back. That is basically teacher forcing.

The purpose of the repo is to show a keras architecture that inputs and outputs variable length sequences and only requires a matrix the size of the largest input output combination using symbolic loops.

It encodes a sequence of characters and decodes a sequence of characters. The results look correct but if you have a good test set you are working on that would be better. Sounds like people are interested in QA and dialog?

Cheers

**Comment by unrealwill** [comment]
Mea culpa, I read your code a little quickly, missed your concatenation of input and outputs and jumped to conclusions.
You are indeed doing teacher forcing.

Still I wouldn't call this a Seq2Seq architecture, but more a sequence continuation architecture (or language model architecture).

The whole point of a Seq2Seq architecture is that the 2 Seq can come from totally different spaces.
The second important point of Seq2Seq architecture is that the network has direct connections to the first sequence when decoding the second sequence, so that it's easier for it to not forget the first sequence, (otherwise you are not using the specific structure of a sequence to sequence problem).

You are doing greedy decoding, but what is needed is beam searching, and you will have a hard time to include it.

It's probably clear by now, I don't like your "architecture" at all. It's confusing, complex, hard to evolve, hard to stack, error prone. Your layer doesn't respect keras "spirit", it is not stackable (composable) with other layer in a way which makes any sense (i.e. it doesn't respect the input in, output out principle). I really hope your code don't find its way into Keras, for it'll be of negative value for lots of beginners though the code has some nice ideas.

**Comment by bstriner** [comment]
@unrealwill Wow. Even github has trolls. Closing an unproductive thread.

The point was that you can do symbolic loops just fine with different sequence lengths if you use a mask for the loss function. Progress works by experimenting and trying different things.

As I said, if you want to do something specific like repeating the sequence and using different decoders, that is something you can do with that architecture. The fact that a model isn't exactly the one you wanted doesn't mean it isn't a model.

I made something overnight that encodes words and outputs predicted next words. You type "mark" it says "antony". This is not intended to be a final architecture, which is why I mentioned rewriting it with recurrentshop, and of course it is not documented or anything.

I make plenty of contributions to keras and I heavily document and test them. That is why I put this in a separate repository and linked to it.

Learn some manners and to appreciate things for what they are instead of just lashing out.
